src/galileo/protect.py

```diff
@@ -0,0 +1,106 @@
+    ) -> Optional[Union[Response, HTTPValidationError]]:
+        request = Request(
+            payload=payload,
+            prioritized_rulesets=prioritized_rulesets or [],
+            project_id=str(project_id) if project_id is not None else None,
+            project_name=project_name,
+            stage_id=str(stage_id) if stage_id is not None else None,
+            stage_name=stage_name,
+            stage_version=stage_version,
+            timeout=timeout,
+            metadata=metadata,
+            headers=headers,
+        )
+        request_dict = request.model_dump(mode="json")
+        request_dict["prioritized_rulesets"] = request_dict.pop("rulesets", [])
+        body = APIRequest.from_dict(request_dict)
+
+        response: Optional[Union[APIResponse, HTTPValidationError]] = invoke_protect_invoke_post.sync(
+            client=self.client, body=body
+        )
+
+        if isinstance(response, APIResponse):
+            return Response.model_validate(response.to_dict())
+        return response
+
+
+def invoke(
+    payload: Payload,
+    prioritized_rulesets: Optional[Sequence[Ruleset]] = None,
+    project_id: Optional[UUID4] = None,
+    project_name: Optional[str] = None,
+    stage_id: Optional[UUID4] = None,
+    stage_name: Optional[str] = None,
+    stage_version: Optional[int] = None,
+    timeout: float = TIMEOUT_SECS,
+    metadata: Optional[dict[str, str]] = None,
+    headers: Optional[dict[str, str]] = None,
+) -> Optional[Union[Response, HTTPValidationError]]:
+    """Invoke Protect with the given payload.
+
+    If using the local stage, the prioritized rulesets should be provided to ensure the
+    correct rulesets are used for processing. If using a central stage, the rulesets
+    will be fetched from the existing stage definition.
+
+    Project ID and stage name, or stage ID should be provided for all invocations.
+
+    Args:
+        payload: Payload to be processed.
+        prioritized_rulesets: Prioritized rulesets to be used for processing.
+            These should only be provided if using a local stage. Defaults to an
+            empty list if None.
+        project_id: ID of the project.
+        project_name: Name of the project.
+        stage_id: ID of the stage.
+        stage_name: Name of the stage.
+        stage_version: Version of the stage.
+        timeout: Timeout for the request in seconds. Defaults to TIMEOUT_SECS.
+        metadata: Metadata to be added when responding.
+        headers: Headers to be added to the response.
+
+    Returns:
+        Protect invoke results.
+    """
+    return Protect().invoke(
+        payload=payload,
+        prioritized_rulesets=prioritized_rulesets,
+        project_id=project_id,
+        project_name=project_name,
+        stage_id=stage_id,
+        stage_name=stage_name,
+        stage_version=stage_version,
+        timeout=timeout,
+        metadata=metadata,
+        headers=headers,
+    )
```

Comment:
No async?

Reply:
It doesn't seem like we have async methods in the client for any of the other calls so hesitated adding one here.

Reply:
We'll need async for Protect. Most of our customers use the async variants.

Reply:
Cool beans, I added `async invoke` to Protect and `async run` to the Protect Tool following the pattern in the `galileo-protect` repo.